### DISCUSSION TITLE
Handling AuxVariable and Problem compatible with multiprocessing/threading

### DISCUSSION THREAD

**Discussion by bielsnohr** [idaholab/moose#19146] [Q&A General] [answered]
I am in the process of creating a MultiApp to call an external program, and the SubApp (contains a class inheriting from ExternalProblem) is meant to both run the external program and then read from the HDF5 file that the program generates. I am currently working on the reading from the HDF5. Some values from this HDF5 are read in, used in some calculations, and then the result is to be stored in an AuxVariable that is passed up to the MainApp. I have a few questions in relation to syncing the results to this AuxVariable.

I want to check that the AuxVariable is present and of the correct family and order, which I do with the following at the top of the `syncSolutions()` class method:
```c++
  // _problem_system is initialised to getAuxiliarySystem() in the class constructor
  if (!_problem_system.hasVariable(_sync_to_var_name))
  {
    throw MooseException("AuxVariable " + _sync_to_var_name +
                         " from input file is not present in the AuxiliarySystem. Make sure you "
                         "have declared the AuxVariables block.");
  }
  MooseVariableFieldBase & fi_heat_fluxes =
      _problem_system.getVariable(_restartable_tid, _sync_to_var_name);

  if (fi_heat_fluxes.isNodal() || fi_heat_fluxes.feType().order != 1)
  {
    throw MooseException("MooseVariable passed to AscotProblem is nodal or of order >1. It must be "
                         "elemental and order 1.");
  }
```
There are two things that bother me about this. First, the use of `MooseVariableFieldBase` which my IDE says is the only valid option. Is that really the correct type for an AuxVariable? Second, the need to pass a thread ID to the `getVariable()` method. I am currently passing `_restartable_tid`, but I am not sure if this is really what I want? This makes me suspect that I am only getting a portion of the AuxVariable, and so I will need to take that into consideration when getting the values calculated from the HDF5 file (indexed by mesh element).

The latter question exposed a confusion in my understanding of how MOOSE handles multiprocessing and multithreading. I understand the that underlying libMesh System and Mesh are handled in a distributed manner, but I don't know whether the the MOOSE class objects are handled in the same way. In other words, is there a `ExternalProblem`-based object that lives in each process/thread, such that when a call is made to `ExternalProblemObject.syncSolutions()` this is executed on each process/thread?

**Comment by GiudGiud** [comment, marked as answer]
Hello

For your first question:
MooseVariableFieldBase is pretty high up in the inheritance chain. If you know it'll be a FE variable you can likely use MooseVariableFEBase. But you can also just write auto and let the compiler figure it out

For your second question:
There is an ExternalProblem on each process. Thread-wise I dont know. I suspect not. Someone needs to write a comprehensive threading documentation in MOOSE. Do you really need to use threads though? Distributed memory is usually better in MOOSE. 

Guillaume

**Reply by bielsnohr** [reply]
> But you can also just write auto and let the compiler figure it out

Yes, I realised this whilst looking at some other code in the MOOSE library after I posted. `auto` is my friend I always forget :stuck_out_tongue: 

> There is an ExternalProblem on each process.

From looking as some other programs, this is what I was beginning to suspect. Thank you.

> Do you really need to use threads though? Distributed memory is usually better in MOOSE.

Nope, don't need threads particularly, just wondering what I should pass to get the variable. @friedmud 's answer clears that up.

**Comment by friedmud** [comment]
Just pass `0` for the thread ID in this case.

The `ExternalProblem` class is not "threaded" (meaning that only one instance of it exists - no matter how many threads you are using).  Therefore it's "on thread 0".

**Reply by bielsnohr** [reply]
Thank you, that clears my thread ID question up.
